src/index.ts

```diff
@@ -0,0 +1,208 @@
+  readonly fragments: string[];
+  readonly sources: string[];
+  readonly dependencies: string[];
+
+  readonly allFragments: string[];
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * read that file. If no file exists, null is returned. If the file exists
+   * but there is a problem reading or parsing it, throw an exception.
+   */
+  static loadOptionsFromFile(filepath: string): ProjectOptions {
+    let configParsed: ProjectOptions;
+    try {
+      const configContent = fs.readFileSync(filepath, 'utf-8');
+      configParsed = JSON.parse(configContent);
+    } catch (error) {
+      // swallow "not found" errors because they are so common / expected
+      if (error.code === 'ENOENT') {
+        logger.debug('no polymer config file found', {file: filepath});
+        return null;
+      }
+      // otherwise, throw an exception
+      throw error;
+    }
+
+    return configParsed;
+  }
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * return a new ProjectConfig instance created with those options.
+   */
+  static loadConfigFromFile(filepath: string): ProjectConfig {
+    let configParsed = ProjectConfig.loadOptionsFromFile(filepath);
+    if (!configParsed) {
+      return null;
+    }
+    return new ProjectConfig(configParsed);
+  }
+
+  /**
+   * constructor - given a ProjectOptions object, create the correct project
+   * configuration for those options. This involves setting the correct
+   * defaults, validating options, warning on deprecated options, and
+   * calculating some additional properties.
+   */
+  constructor(options: ProjectOptions) {
+    options = options || {};
+    options = fixDeprecatedOptions(options);
+
+    /**
+     * root
+     */
+    this.root = process.cwd();
+    if (options.root) {
+      this.root = path.resolve(this.root, options.root);
+    }
+
+    /**
+     * entrypoint
+     */
+    if (options.entrypoint) {
+      this.entrypoint = path.resolve(this.root, options.entrypoint);
+    } else {
+      this.entrypoint = path.resolve(this.root, 'index.html');
+    }
+
+    /**
+     * shell
+     */
+    if (options.shell) {
+      this.shell = path.resolve(this.root, options.shell);
+    }
+
+    /**
+     * fragments
+     */
+    if (options.fragments) {
+      this.fragments = options.fragments.map((e) => path.resolve(this.root, e));
+    }
+
+    /**
+     * dependencies
+     */
+    this.dependencies = (options.dependencies || [])
+        .map((glob) => resolveGlob(this.root, glob));
+
+    /**
+     * sources
+     */
+    this.sources = (options.sources || [])
+        .map((glob) => resolveGlob(this.root, glob));
+
+    this.sources.push(this.entrypoint);
+    if (this.shell) {
+      this.sources.push(this.shell);
+    }
+    if (this.fragments) {
+      this.sources = this.sources.concat(this.fragments);
+    }
+
+    /**
+     * allFragments
+     */
+    this.allFragments = [];
```

Comment:
@justinfagnani What about calling this `fragments`, and changing the polymer.json field to `extraFragments` along with `extraSources` & `extraDependencies`? That's really what it's defining right now, and it would clear up https://github.com/Polymer/polymer-build/pull/43#pullrequestreview-384597 as well.